clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

```diff
@@ -690,9 +712,14 @@ class CIRFuncLowering : public mlir::OpConversionPattern<mlir::cir::FuncOp> {
       Loc = FusedLoc.getLocations()[0];
     }
     assert(Loc.isa<mlir::FileLineColLoc>() && "expected single location here");
+
     auto linkage = convertLinkage(op.getLinkage());
-    auto fn = rewriter.create<mlir::LLVM::LLVMFuncOp>(Loc, op.getName(),
-                                                      llvmFnTy, linkage);
+    SmallVector<mlir::NamedAttribute, 4> attributes;
+    filterFuncAttributes(op, /*filterArgAndResAttrs=*/false, attributes);
```

Comment:
What does this actually end up filtering? Your test case doesn't seem to remove anything.

Reply:
So when a CIR funcop is created, all attributes including function attributes, linkage, parameter attributes .... are merged into the underlying attribute array (i.e, https://github.com/llvm/clangir/blob/0d02bda5c5314105f702ad82490b898adb3c63be/clang/lib/CIR/Dialect/IR/CIRDialect.cpp#L1349). The process is repeated when the CIR funcop is converted to MLIR LLVM dialect funcop (
https://github.com/llvm/clangir/blob/c1fa60b4cde512964544ab66404dea79dbc5dcb4/mlir/lib/Dialect/Func/IR/FuncOps.cpp#L257). Without filtering, the already-merged attrs array will be merged again with those attributes namely linkage type, parameter attrs..., causing duplicates.